src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs

```diff
@@ -0,0 +1,360 @@
+                    ref certHandleSize))
+                {
+                    int lastError = Marshal.GetLastWin32Error();
+                    throw WinHttpException.CreateExceptionUsingError(lastError);
+                }
+                
+                // Create a managed wrapper around the certificate handle. Since this results in duplicating
+                // the handle, we will close the original handle after creating the wrapper.
+                var serverCertificate = new X509Certificate2(certHandle);
+                Interop.Crypt32.CertFreeCertificateContext(certHandle);
+
+                X509Chain chain = null;
+                SslPolicyErrors sslPolicyErrors;
+
+                try
+                {
+                    WinHttpCertificateHelper.BuildChain(
+                        serverCertificate,
+                        state.RequestMessage.RequestUri.Host,
+                        state.CheckCertificateRevocationList,
+                        out chain,
+                        out sslPolicyErrors);
+
+                    bool result = state.ServerCertificateValidationCallback(
+                        state.RequestMessage,
+                        serverCertificate,
+                        chain,
+                        sslPolicyErrors);
+                    if (!result)
+                    {
+                        throw WinHttpException.CreateExceptionUsingError(
+                            (int)Interop.WinHttp.ERROR_WINHTTP_SECURE_FAILURE);
+                    }
+                }
+                finally
+                {
+                    if (chain != null)
+                    {
+                        chain.Dispose();
+                    }
+                }
+            }
+        }
+
+        private static void OnRequestError(WinHttpRequestState state, Interop.WinHttp.WINHTTP_ASYNC_RESULT asyncResult)
+        {
+            WinHttpTraceHelper.TraceAsyncError("OnRequestError", asyncResult);
+            
+            Debug.Assert(state != null, "OnRequestError: state is null");
+
+            var innerException = WinHttpException.CreateExceptionUsingError((int)asyncResult.dwError);
+
+            switch ((uint)asyncResult.dwResult.ToInt32())
+            {
+                case Interop.WinHttp.API_SEND_REQUEST:
+                    state.TcsSendRequest.TrySetException(innerException);
+                    break;
+                    
+                case Interop.WinHttp.API_RECEIVE_RESPONSE:
+                    if (asyncResult.dwError == Interop.WinHttp.ERROR_WINHTTP_RESEND_REQUEST)
+                    {
+                        state.RetryRequest = true;
+                        state.TcsReceiveResponseHeaders.TrySetResult(false);
+                    }
+                    else if (asyncResult.dwError == Interop.WinHttp.ERROR_WINHTTP_CLIENT_AUTH_CERT_NEEDED)
+                    {
+                        // WinHttp will automatically drop any client SSL certificates that we
```

Comment:
I'm not sure I understand the comment: 
> WinHttp will automatically drop any client SSL certificates

vs

> But we need to tell WinHttp we don't have any certificate.

I notice that we still clear it. Is the comment out of sync?